File: ZOJ/1331.py
#Problem ID: 1331
#Submit Time: 2013-02-27 11:40:11
#Run Time: 1160
#Run Memory: 316
#ZOJ User: calvinxiao

cube = [a*a*a for a in xrange(201)]
mp = {}
for i in xrange(201):
    mp[cube[i]] = i

ans = []
for b in xrange(2, 201):
    for c in xrange(b+1, 201):
        for d in xrange(c+1, 201):
            t = cube[b] + cube[c] + cube[d]
            if t in mp:
                #print t, int(t**0.3333+0.5)
                ans.append([mp[t], b, c, d])
ans.sort()
for i in xrange(len(ans)):
    print "Cube = %d, Triple = (%d,%d,%d)" % (ans[i][0],ans[i][1],ans[i][2],ans[i][3])

File: ZOJ/1331.cpp
/*
#Problem ID: 1331
#Submit Time: 2013-02-27 07:44:44
#Run Time: 70
#Run Memory: 31908
#ZOJ User: calvinxiao
*/

#include <iostream>
#include <cstdio>
#include <set>
#include <vector>
#include <map>
#include <stack>
#include <queue>
#include <cmath>
#include <algorithm>
#include <cstring>
#include <string>
#include <sstream>

#define REP(i, n) for(int (i) = 0; (i) < (int)(n); ++(i))
#define REPN(i, a, n) for(int (i) = (int)(a); (i) <= (n); ++(i))
#define FORE(itr, e) for(typeof(e.begin) itr = e.begin(); itr != e.end(); itr++)

using namespace std;

const int BIG = (int)1e9 + 7;

const int MAX = 201*201*201;
int cube[201], a, b, c, d, p, product[MAX];
pair<int, vector<int> > pa[256];

int main()
{
	//freopen("0.in", "r", stdin);
	for (a = 2; a <= 200; ++a) {
        cube[a] = a*a*a;
        product[cube[a]] = a;
    }
        
    int cnt = 0;
    for (int i = 2; i <= 200; ++i) {
        for (int j = i+1; j <= 200; ++j) {
            for (int k = j+1; k <= 200; ++k) {
                p = cube[i] + cube[j] + cube[k];
                if (p > MAX)
                    continue;
                if (product[p]) {
                    vector<int> v; v.push_back(i); v.push_back(j); v.push_back(k);
                    pa[cnt++] = make_pair(product[p], v);
                }
            }
        }
    }
    sort(pa, pa + cnt);
    for (int i = 0; i < cnt; ++i) {
        printf("Cube = %d, Triple = (%d,%d,%d)\n", pa[i].first, pa[i].second[0], pa[i].second[1], pa[i].second[2]);
    }

}
